File: beatmap_preview/parser.py
from __future__ import annotations

from pathlib import Path

from .errors import PreviewError
from .models import (
    Beatmap,
    BreakPeriod,
    CatchHitObject,
    HitObject,
    ManiaHitObject,
    StandardHitObject,
    TaikoHitObject,
    TimingPoint,
)


def parse_beatmap(beatmap_path: Path) -> Beatmap:
    try:
        content = beatmap_path.read_text(encoding="utf-8-sig")
        sections = _split_sections(content)

        # .osu 文件把 Metadata、Difficulty 等内容拆在不同 section 中；后续解析都基于这些分组。
        metadata = _parse_key_value_section(sections, "Metadata") if "Metadata" in sections else _default_metadata()
        difficulty = _parse_key_value_section(sections, "Difficulty")
        general = _parse_key_value_section(sections, "General") if "General" in sections else {"Mode": "0"}
        general["FormatVersion"] = str(_parse_format_version(content))
        timing_points = _parse_timing_points(sections)
        break_periods = _parse_break_periods(sections)
        mode = int(general["Mode"])

        # 各 mode 的 HitObjects 字段含义不同，先分派到专门解析函数再统一排序。
        if mode == 0:
            hit_objects = _parse_standard_hit_objects(sections, difficulty, timing_points)
        elif mode == 1:
            hit_objects = _parse_taiko_hit_objects(sections, difficulty, timing_points)
        elif mode == 2:
            hit_objects = _parse_catch_hit_objects(sections, difficulty, timing_points)
        elif mode == 3:
            hit_objects = _parse_mania_hit_objects(sections, difficulty, timing_points)
        else:
            raise PreviewError(f"Unsupported beatmap mode: {mode}")

        return Beatmap(
            metadata=metadata,
            difficulty=difficulty,
            general=general,
            timing_points=timing_points,
            hit_objects=hit_objects,
            break_periods=break_periods,
        )
    except PreviewError:
        raise
    except (OSError, UnicodeError, KeyError, ValueError, IndexError, ZeroDivisionError) as exc:
        raise PreviewError("Failed to parse beatmap.") from exc


def _split_sections(content: str) -> dict[str, list[str]]:
    sections: dict[str, list[str]] = {}
    current_section = ""

    for raw_line in content.splitlines():
        line = raw_line.strip()
        if not line or line.startswith("//"):
            continue

        if line.startswith("[") and line.endswith("]"):
            current_section = line[1:-1]
            sections[current_section] = []
            continue

        if not current_section:
            continue
        sections[current_section].append(line)

    return sections


def _parse_format_version(content: str) -> int:
    lines = content.splitlines()
    first_line = lines[0].strip() if lines else ""
    if first_line.startswith("osu file format v"):
        return int(first_line.rsplit("v", 1)[1])
    return 14


def _parse_key_value_section(sections: dict[str, list[str]], section_name: str) -> dict[str, str]:
    values: dict[str, str] = {}
    for line in sections[section_name]:
        if ":" not in line:
            continue
        key, value = line.split(":", 1)
        values[key.strip()] = value.strip()

    return values


def _default_metadata() -> dict[str, str]:
    return {
        "Title": "Unknown",
        "TitleUnicode": "",
        "Artist": "Unknown",
        "ArtistUnicode": "",
        "Creator": "Unknown",
        "Version": "Unknown",
    }


def _parse_timing_points(sections: dict[str, list[str]]) -> list[TimingPoint]:
    timing_points: list[TimingPoint] = []
    for line in sections["TimingPoints"]:
        parts = [part.strip() for part in line.split(",")]
        if len(parts) < 2:
            continue

        meter = int(parts[2]) if len(parts) > 2 and parts[2] else 4
        if meter <= 0:
            meter = 4
        # 第 7 列为 1 或缺失时是红线 BPM；0 则是 inherited velocity。
        uninherited = len(parts) < 7 or parts[6] == "1"
        # 第 8 列 effects 的最低位表示 kiai 是否开启。
        effects = int(parts[7]) if len(parts) > 7 and parts[7] else 0
        timing_points.append(
            TimingPoint(
                time=float(parts[0]),
                beat_length=float(parts[1]),
                meter=meter,
                uninherited=uninherited,
                kiai_mode=bool(effects & 1),
            )
        )

    # 同一时间点的红线 / 绿线顺序会影响后续 timing 计算，必须保留文件里的原始顺序。
    return sorted(timing_points, key=lambda point: point.time)


def _parse_break_periods(sections: dict[str, list[str]]) -> list[BreakPeriod]:
    if "Events" not in sections:
        return []

    break_periods: list[BreakPeriod] = []
    for line in sections["Events"]:
        parts = [part.strip() for part in line.split(",")]
        if len(parts) < 3 or parts[0] != "2":
            continue

        start_time = int(float(parts[1]))
        end_time = int(float(parts[2]))
        if end_time > start_time:
            break_periods.append(BreakPeriod(start_time=start_time, end_time=end_time))

    return break_periods


def _parse_standard_hit_objects(
    sections: dict[str, list[str]],
    difficulty: dict[str, str],
    timing_points: list[TimingPoint],
) -> list[HitObject]:
    hit_objects: list[HitObject] = []
    for line in sections["HitObjects"]:
        parts = [part.strip() for part in line.split(",")]
        if len(parts) < 5:
            continue

        x = int(parts[0])
        y = int(parts[1])
        start_time = int(parts[2])
        hit_type = int(parts[3])
        hitsound = int(parts[4])
        end_time = _parse_object_end_time(parts, start_time, hit_type, difficulty, timing_points)
        new_combo = bool(hit_type & 4)
        combo_offset = (hit_type & 112) >> 4
        slider_type = None
        slider_points: tuple[tuple[int, int], ...] = ()
        slider_repeats = 1
        slider_pixel_length = 0.0
        slider_edge_hitsounds: tuple[int, ...] = ()

        if hit_type & 2:
            # Slider 专有字段从第 6 列开始：类型、控制点、重复次数和像素长度。
            slider_parts = parts[5].split("|")
            slider_type = slider_parts[0]
            slider_points = tuple(
                (int(point.split(":", 1)[0]), int(point.split(":", 1)[1]))
                for point in slider_parts[1:]
            )
            slider_repeats = int(parts[6])
            slider_pixel_length = float(parts[7])
            if len(parts) > 8 and parts[8]:
                slider_edge_hitsounds = tuple(int(value) for value in parts[8].split("|") if value)

        hit_objects.append(
            StandardHitObject(
                x=x,
                y=y,
                start_time=start_time,
                end_time=end_time,
                hit_type=hit_type,
                hitsound=hitsound,
                new_combo=new_combo,
                combo_offset=combo_offset,
                slider_type=slider_type,
                slider_points=slider_points,
                slider_repeats=slider_repeats,
                slider_pixel_length=slider_pixel_length,
                slider_edge_hitsounds=slider_edge_hitsounds,
            )
        )

    return _sort_hit_objects(hit_objects)


def _parse_taiko_hit_objects(
    sections: dict[str, list[str]],
    difficulty: dict[str, str],
    timing_points: list[TimingPoint],
) -> list[HitObject]:
    hit_objects: list[HitObject] = []
    for line in sections["HitObjects"]:
        parts = [part.strip() for part in line.split(",")]
        if len(parts) < 5:
            continue

        start_time = int(parts[2])
        hit_type = int(parts[3])
        hitsound = int(parts[4])
        end_time = _parse_object_end_time(parts, start_time, hit_type, difficulty, timing_points)

        hit_objects.append(
            TaikoHitObject(
                start_time=start_time,
                end_time=end_time,
                hit_type=hit_type,
                hitsound=hitsound,
            )
        )

    return _sort_hit_objects(hit_objects)


def _parse_catch_hit_objects(
    sections: dict[str, list[str]],
    difficulty: dict[str, str],
    timing_points: list[TimingPoint],
) -> list[HitObject]:
    hit_objects: list[HitObject] = []
    for line in sections["HitObjects"]:
        parts = [part.strip() for part in line.split(",")]
        if len(parts) < 5:
            continue

        x = int(parts[0])
        y = int(parts[1])
        start_time = int(parts[2])
        hit_type = int(parts[3])
        end_time = _parse_object_end_time(parts, start_time, hit_type, difficulty, timing_points)
        new_combo = bool(hit_type & 4)
        combo_offset = (hit_type & 112) >> 4
        slider_type = None
        slider_points: tuple[tuple[int, int], ...] = ()
        slider_repeats = 1
        slider_pixel_length = 0.0

        if hit_type & 2:
            slider_parts = parts[5].split("|")
            slider_type = slider_parts[0]
            slider_points = tuple(
                (int(point.split(":", 1)[0]), int(point.split(":", 1)[1]))
                for point in slider_parts[1:]
            )
            slider_repeats = int(parts[6])
            slider_pixel_length = float(parts[7])

        hit_objects.append(
            CatchHitObject(
                x=x,
                y=y,
                start_time=start_time,
                end_time=end_time,
                hit_type=hit_type,
                new_combo=new_combo,
                combo_offset=combo_offset,
                slider_type=slider_type,
                slider_points=slider_points,
                slider_repeats=slider_repeats,
                slider_pixel_length=slider_pixel_length,
            )
        )

    return _sort_hit_objects(hit_objects)


def _parse_mania_hit_objects(
    sections: dict[str, list[str]],
    difficulty: dict[str, str],
    timing_points: list[TimingPoint],
) -> list[HitObject]:
    key_count = int(float(difficulty["CircleSize"]))
    hit_objects: list[HitObject] = []
    for line in sections["HitObjects"]:
        parts = [part.strip() for part in line.split(",")]
        if len(parts) < 5:
            continue

        x = int(parts[0])
        start_time = int(parts[2])
        hit_type = int(parts[3])
        # mania 使用 x 坐标按键数等分出 lane，避免 renderer 再依赖原始坐标。
        lane = max(0, min(key_count - 1, x * key_count // 512))
        is_long_note = bool(hit_type & 128)
        end_time = start_time
        if is_long_note:
            end_time = int(parts[5].split(":", 1)[0])

        hit_objects.append(
            ManiaHitObject(
                lane=lane,
                start_time=start_time,
                end_time=end_time,
                is_long_note=is_long_note,
            )
        )

    return _sort_hit_objects(hit_objects)


def _parse_object_end_time(
    parts: list[str],
    start_time: int,
    hit_type: int,
    difficulty: dict[str, str],
    timing_points: list[TimingPoint],
) -> int:
    # osu! hit_type 位标志：8=spinner，2=slider；普通物件结束时间就是开始时间。
    if hit_type & 8:
        return int(parts[5])
    if hit_type & 2:
        return _parse_slider_end_time(parts, start_time, difficulty, timing_points)
    return start_time


def _parse_slider_end_time(
    parts: list[str],
    start_time: int,
    difficulty: dict[str, str],
    timing_points: list[TimingPoint],
) -> int:
    # Slider 时长由当前红线 BPM 和最近的 inherited velocity 共同决定。
    slides = int(parts[6])
    pixel_length = float(parts[7])
    slider_multiplier = float(difficulty["SliderMultiplier"])
    beat_length, slider_velocity = _resolve_slider_timing(start_time, timing_points)
    duration = pixel_length / (slider_multiplier * 100 * slider_velocity) * beat_length * slides
    return start_time + round(duration)


def _resolve_slider_timing(start_time: int, timing_points: list[TimingPoint]) -> tuple[float, float]:
    beat_length = timing_points[0].beat_length
    slider_velocity = 1.0

    # 扫到物件开始时间前的最后有效 timing：红线更新 BPM，绿线更新 slider velocity。
    for point in timing_points:
        if point.time > start_time:
            break
        if point.uninherited:
            beat_length = point.beat_length
            slider_velocity = 1.0
        elif point.beat_length < 0:
            slider_velocity = -100 / point.beat_length

    return beat_length, slider_velocity


def _sort_hit_objects(hit_objects: list[HitObject]) -> list[HitObject]:
    return sorted(hit_objects, key=lambda hit_object: (hit_object.start_time, hit_object.end_time))

File: src/parser.rs
use crate::errors::{PreviewError, Result};
use crate::models::*;
use std::collections::HashMap;
use std::path::Path;

pub fn parse_beatmap(path: &Path) -> Result<Beatmap> {
    let bytes = std::fs::read(path).map_err(|_| PreviewError::new("Failed to parse beatmap."))?;
    let content = String::from_utf8_lossy(&bytes);
    let content = content.strip_prefix('\u{feff}').unwrap_or(&content);
    parse_beatmap_str(content).ok_or_else(|| PreviewError::new("Failed to parse beatmap."))
}

fn parse_beatmap_str(content: &str) -> Option<Beatmap> {
    let sections = split_sections(content);

    let metadata = match sections.get("Metadata") {
        Some(lines) => parse_key_value(lines),
        None => default_metadata(),
    };
    let difficulty = parse_key_value(sections.get("Difficulty")?);
    let mut general = match sections.get("General") {
        Some(lines) => parse_key_value(lines),
        None => {
            let mut kv = KvSection::default();
            kv.insert("Mode", "0".to_string());
            kv
        }
    };
    general.insert("FormatVersion", parse_format_version(content).to_string());
    let timing_points = parse_timing_points(sections.get("TimingPoints")?)?;
    let break_periods = parse_break_periods(sections.get("Events"));
    let mode: i32 = general.get("Mode").unwrap_or("0").parse().ok()?;

    let combo_colors = parse_combo_colors(sections.get("Colours"));

    let editor = sections
        .get("Editor")
        .map(|lines| parse_key_value(lines))
        .unwrap_or_default();
    let beat_divisor: i32 = editor.get("BeatDivisor").and_then(|v| v.parse().ok()).unwrap_or(0);

    let hit_lines = sections.get("HitObjects")?;
    let hit_objects = match mode {
        0 => HitObjects::Standard(parse_standard(hit_lines, &difficulty, &timing_points)?),
        1 => HitObjects::Taiko(parse_taiko(hit_lines, &difficulty, &timing_points)?),
        2 => HitObjects::Catch(parse_catch(hit_lines, &difficulty, &timing_points)?),
        3 => HitObjects::Mania(parse_mania(hit_lines, &difficulty)?),
        _ => return None,
    };

    Some(Beatmap {
        metadata,
        difficulty,
        general,
        timing_points,
        hit_objects,
        break_periods,
        combo_colors,
        beat_divisor,
    })
}

/// Parse Combo1..ComboN from the [Colours] section, in numeric order.
fn parse_combo_colors(lines: Option<&Vec<&str>>) -> Vec<[u8; 3]> {
    let Some(lines) = lines else {
        return Vec::new();
    };
    let mut entries: Vec<(u32, [u8; 3])> = Vec::new();
    for line in lines {
        let Some((key, value)) = line.split_once(':') else {
            continue;
        };
        let key = key.trim();
        let Some(num) = key.strip_prefix("Combo") else {
            continue;
        };
        let Ok(index) = num.parse::<u32>() else {
            continue;
        };
        let parts: Vec<&str> = value.split(',').map(|p| p.trim()).collect();
        if parts.len() < 3 {
            continue;
        }
        let (Ok(r), Ok(g), Ok(b)) = (
            parts[0].parse::<u8>(),
            parts[1].parse::<u8>(),
            parts[2].parse::<u8>(),
        ) else {
            continue;
        };
        entries.push((index, [r, g, b]));
    }
    entries.sort_by_key(|e| e.0);
    entries.into_iter().map(|e| e.1).collect()
}

fn split_sections(content: &str) -> HashMap<String, Vec<&str>> {
    let mut sections: HashMap<String, Vec<&str>> = HashMap::new();
    let mut current: Option<String> = None;

    for raw in content.lines() {
        let line = raw.trim();
        if line.is_empty() || line.starts_with("//") {
            continue;
        }
        if line.starts_with('[') && line.ends_with(']') {
            let name = line[1..line.len() - 1].to_string();
            sections.entry(name.clone()).or_default();
            current = Some(name);
            continue;
        }
        if let Some(name) = &current {
            sections.get_mut(name).unwrap().push(line);
        }
    }
    sections
}

fn parse_format_version(content: &str) -> i32 {
    let first = content.lines().next().unwrap_or("").trim();
    if let Some(rest) = first.strip_prefix("osu file format v") {
        if let Ok(v) = rest.parse() {
            return v;
        }
    }
    14
}

fn parse_key_value(lines: &[&str]) -> KvSection {
    let mut kv = KvSection::default();
    for line in lines {
        if let Some((key, value)) = line.split_once(':') {
            kv.insert(key.trim(), value.trim().to_string());
        }
    }
    kv
}

fn default_metadata() -> KvSection {
    let mut kv = KvSection::default();
    kv.insert("Title", "Unknown".into());
    kv.insert("TitleUnicode", "".into());
    kv.insert("Artist", "Unknown".into());
    kv.insert("ArtistUnicode", "".into());
    kv.insert("Creator", "Unknown".into());
    kv.insert("Version", "Unknown".into());
    kv
}

fn parse_timing_points(lines: &[&str]) -> Option<Vec<TimingPoint>> {
    let mut points: Vec<TimingPoint> = Vec::new();
    for line in lines {
        let parts: Vec<&str> = line.split(',').map(|p| p.trim()).collect();
        if parts.len() < 2 {
            continue;
        }
        let mut meter = if parts.len() > 2 && !parts[2].is_empty() {
            parts[2].parse::<i32>().ok()?
        } else {
            4
        };
        if meter <= 0 {
            meter = 4;
        }
        let uninherited = parts.len() < 7 || parts[6] == "1";
        let effects = if parts.len() > 7 && !parts[7].is_empty() {
            parts[7].parse::<i32>().ok()?
        } else {
            0
        };
        points.push(TimingPoint {
            time: parts[0].parse().ok()?,
            beat_length: parts[1].parse().ok()?,
            meter,
            uninherited,
            kiai_mode: effects & 1 != 0,
        });
    }
    // Stable sort keeps file order for equal times (red/green at same time).
    points.sort_by(|a, b| a.time.partial_cmp(&b.time).unwrap());
    if points.is_empty() {
        return None;
    }
    Some(points)
}

fn parse_break_periods(lines: Option<&Vec<&str>>) -> Vec<BreakPeriod> {
    let Some(lines) = lines else {
        return Vec::new();
    };
    let mut breaks = Vec::new();
    for line in lines {
        let parts: Vec<&str> = line.split(',').map(|p| p.trim()).collect();
        if parts.len() < 3 || parts[0] != "2" {
            continue;
        }
        let (Ok(s), Ok(e)) = (parts[1].parse::<f64>(), parts[2].parse::<f64>()) else {
            continue;
        };
        let (start_time, end_time) = (s as i64, e as i64);
        if end_time > start_time {
            breaks.push(BreakPeriod {
                start_time,
                end_time,
            });
        }
    }
    breaks
}

struct SliderFields {
    slider_type: String,
    points: Vec<(i32, i32)>,
    repeats: i32,
    pixel_length: f64,
    edge_hitsounds: Vec<i32>,
}

fn parse_slider_fields(parts: &[&str]) -> Option<SliderFields> {
    let mut slider_parts = parts[5].split('|');
    let slider_type = slider_parts.next()?.to_string();
    let mut points = Vec::new();
    for p in slider_parts {
        let (x, y) = p.split_once(':')?;
        points.push((x.parse().ok()?, y.parse().ok()?));
    }
    let repeats: i32 = parts.get(6)?.parse().ok()?;
    let pixel_length: f64 = parts.get(7)?.parse().ok()?;
    let mut edge_hitsounds = Vec::new();
    if let Some(eh) = parts.get(8) {
        if !eh.is_empty() {
            for v in eh.split('|') {
                if !v.is_empty() {
                    edge_hitsounds.push(v.parse().ok()?);
                }
            }
        }
    }
    Some(SliderFields {
        slider_type,
        points,
        repeats,
        pixel_length,
        edge_hitsounds,
    })
}

fn parse_standard(
    lines: &[&str],
    difficulty: &KvSection,
    timing_points: &[TimingPoint],
) -> Option<Vec<StandardHitObject>> {
    let mut objects = Vec::with_capacity(lines.len());
    for line in lines {
        let parts: Vec<&str> = line.split(',').map(|p| p.trim()).collect();
        if parts.len() < 5 {
            continue;
        }
        let x: i32 = parts[0].parse::<f64>().ok()? as i32;
        let y: i32 = parts[1].parse::<f64>().ok()? as i32;
        let start_time: i64 = parts[2].parse().ok()?;
        let hit_type: i32 = parts[3].parse().ok()?;
        let hitsound: i32 = parts[4].parse().ok()?;
        let end_time = parse_end_time(&parts, start_time, hit_type, difficulty, timing_points)?;

        let mut obj = StandardHitObject {
            x,
            y,
            start_time,
            end_time,
            hit_type,
            hitsound,
            new_combo: hit_type & 4 != 0,
            combo_offset: (hit_type & 112) >> 4,
            ..Default::default()
        };
        if hit_type & 2 != 0 {
            let sf = parse_slider_fields(&parts)?;
            obj.slider_type = Some(sf.slider_type);
            obj.slider_points = sf.points;
            obj.slider_repeats = sf.repeats;
            obj.slider_pixel_length = sf.pixel_length;
            obj.slider_edge_hitsounds = sf.edge_hitsounds;
        }
        objects.push(obj);
    }
    objects.sort_by_key(|o| (o.start_time, o.end_time));
    Some(objects)
}

fn parse_taiko(
    lines: &[&str],
    difficulty: &KvSection,
    timing_points: &[TimingPoint],
) -> Option<Vec<TaikoHitObject>> {
    let mut objects = Vec::with_capacity(lines.len());
    for line in lines {
        let parts: Vec<&str> = line.split(',').map(|p| p.trim()).collect();
        if parts.len() < 5 {
            continue;
        }
        let start_time: i64 = parts[2].parse().ok()?;
        let hit_type: i32 = parts[3].parse().ok()?;
        let hitsound: i32 = parts[4].parse().ok()?;
        let end_time = parse_end_time(&parts, start_time, hit_type, difficulty, timing_points)?;
        objects.push(TaikoHitObject {
            start_time,
            end_time,
            hit_type,
            hitsound,
        });
    }
    objects.sort_by_key(|o| (o.start_time, o.end_time));
    Some(objects)
}

fn parse_catch(
    lines: &[&str],
    difficulty: &KvSection,
    timing_points: &[TimingPoint],
) -> Option<Vec<CatchHitObject>> {
    let mut objects = Vec::with_capacity(lines.len());
    for line in lines {
        let parts: Vec<&str> = line.split(',').map(|p| p.trim()).collect();
        if parts.len() < 5 {
            continue;
        }
        let x: i32 = parts[0].parse::<f64>().ok()? as i32;
        let y: i32 = parts[1].parse::<f64>().ok()? as i32;
        let start_time: i64 = parts[2].parse().ok()?;
        let hit_type: i32 = parts[3].parse().ok()?;
        let end_time = parse_end_time(&parts, start_time, hit_type, difficulty, timing_points)?;

        let mut obj = CatchHitObject {
            x,
            y,
            start_time,
            end_time,
            hit_type,
            new_combo: hit_type & 4 != 0,
            combo_offset: (hit_type & 112) >> 4,
            slider_type: None,
            slider_points: Vec::new(),
            slider_repeats: 1,
            slider_pixel_length: 0.0,
        };
        if hit_type & 2 != 0 {
            let sf = parse_slider_fields(&parts)?;
            obj.slider_type = Some(sf.slider_type);
            obj.slider_points = sf.points;
            obj.slider_repeats = sf.repeats;
            obj.slider_pixel_length = sf.pixel_length;
        }
        objects.push(obj);
    }
    objects.sort_by_key(|o| (o.start_time, o.end_time));
    Some(objects)
}

fn parse_mania(lines: &[&str], difficulty: &KvSection) -> Option<Vec<ManiaHitObject>> {
    let key_count = difficulty.get_f64("CircleSize")? as i64;
    let mut objects = Vec::with_capacity(lines.len());
    for line in lines {
        let parts: Vec<&str> = line.split(',').map(|p| p.trim()).collect();
        if parts.len() < 5 {
            continue;
        }
        let x: i64 = parts[0].parse::<f64>().ok()? as i64;
        let start_time: i64 = parts[2].parse().ok()?;
        let hit_type: i32 = parts[3].parse().ok()?;
        let lane = (x * key_count).div_euclid(512).clamp(0, key_count - 1) as i32;
        let is_long_note = hit_type & 128 != 0;
        let mut end_time = start_time;
        if is_long_note {
            let head = parts.get(5)?.split(':').next()?;
            end_time = head.parse().ok()?;
        }
        objects.push(ManiaHitObject {
            lane,
            start_time,
            end_time,
            is_long_note,
        });
    }
    objects.sort_by_key(|o| (o.start_time, o.end_time));
    Some(objects)
}

fn parse_end_time(
    parts: &[&str],
    start_time: i64,
    hit_type: i32,
    difficulty: &KvSection,
    timing_points: &[TimingPoint],
) -> Option<i64> {
    if hit_type & 8 != 0 {
        return parts.get(5)?.parse::<f64>().ok().map(|v| v as i64);
    }
    if hit_type & 2 != 0 {
        return parse_slider_end_time(parts, start_time, difficulty, timing_points);
    }
    Some(start_time)
}

fn parse_slider_end_time(
    parts: &[&str],
    start_time: i64,
    difficulty: &KvSection,
    timing_points: &[TimingPoint],
) -> Option<i64> {
    let slides: f64 = parts.get(6)?.parse::<i32>().ok()? as f64;
    let pixel_length: f64 = parts.get(7)?.parse().ok()?;
    let slider_multiplier = difficulty.get_f64("SliderMultiplier")?;
    let (beat_length, slider_velocity) = resolve_slider_timing(start_time, timing_points);
    let duration = pixel_length / (slider_multiplier * 100.0 * slider_velocity) * beat_length * slides;
    Some(start_time + round_half_even(duration))
}

pub fn resolve_slider_timing(start_time: i64, timing_points: &[TimingPoint]) -> (f64, f64) {
    let mut beat_length = timing_points[0].beat_length;
    let mut slider_velocity = 1.0;
    for point in timing_points {
        if point.time > start_time as f64 {
            break;
        }
        if point.uninherited {
            beat_length = point.beat_length;
            slider_velocity = 1.0;
        } else if point.beat_length < 0.0 {
            slider_velocity = -100.0 / point.beat_length;
        }
    }
    (beat_length, slider_velocity)
}

// Python's round() = banker's rounding.
pub fn round_half_even(v: f64) -> i64 {
    let floor = v.floor();
    let diff = v - floor;
    if diff > 0.5 {
        floor as i64 + 1
    } else if diff < 0.5 {
        floor as i64
    } else {
        let f = floor as i64;
        if f % 2 == 0 {
            f
        } else {
            f + 1
        }
    }
}
